docs/Changelog.md

```diff
@@ -1010,6 +1010,249 @@ opset_import {
+            }
+            node {
+              input: "b"
+              input: "b"
+              output: "user_defined_vals"
+              op_type: "Add"
+            }
+            name: "body-net"
+            input {
+              name: "b"
+              type {
+                tensor_type {
+                  elem_type: INT32
+                  shape {
+                  }
+                }
+              }
+            }
+            input {
+              name: "keepgoing"
+              type {
+                tensor_type {
+                  elem_type: BOOL
+                  shape {
+                  }
+                }
+              }
+            }
+            output {
+              name: "b_out"
+              type {
+                tensor_type {
+                  elem_type: INT32
+                  shape {
+                  }
+                }
+              }
+            }
+            output {
+              name: "keepgoing_out"
+              type {
+                tensor_type {
+                  elem_type: BOOL
+                  shape {
+                  }
+                }
+              }
+            }
+            output {
+              name: "user_defined_vals"
+              type {
+                tensor_type {
+                  elem_type: INT32
+                  shape {
+                  }
+                }
+              }
+            }
+          }
+          type: GRAPH
+        }
+      }
+      name: "predict-net"
+  
+    == Sample equivalent C code ==
+  
+      {
+        /* User-defined code (enclosing scope) */
+        int a = 3, b = 6;
+        bool keepgoing = true; // Analogous to input cond
+        /* End user-defined code */
+  
+        /* Implicitly-defined code */
+        int i = 0; // implicitly defined trip count
+        const int max_trip_count = 10; // Analogous to input M
+        int user_defined_vals[]; // Imagine this is resizable
+        /* End implicitly-defined code */
+        do {
+          /* User-defined code (loop body) */
+          int my_local = a + b; // Reading values in the enclosing scope is fine
+          b = a - b; // writes fine if we specify b as a loop-carried dependency
+          keepgoing = my_local > b; // keepgoing is a loop-carried dependency
+          user_defined_vals[i] = b + b;
+          /* End user-defined code */
+  
+          i = i + 1; // implicitly defined trip count
+        } while (i < max_trip_count && keepgoing);
+        // my_local = 123; // Can't do this. my_local was defined in the the body
+      }
+  
+  There are several things of note in this code snippet:
+  
+  1) Reads from values in the enclosing scope (i.e. variable a here) are
+     implicitly allowed.
+  2) Writes to variables in the enclosing scope (i.e. the variables b and
+      keepgoing) MUST have the corresponding values appearing in the loop-carried
+      dependencies, both at the op input and body net input and output.
```

Comment:
I'm not a fan of the read/write nomenclature, since that implies some sort of mutation is going on, but we're SSA, and we're definitely not in the business of having a memory model. As an alternate wording:

1) Values from the enclosing scope (i.e. variable a here) are in scope and can be referenced in the inputs of the loop
2) Any variables which you wish to make available in the enclosing scope (i.e. the variables b and keepgoing) must be declared in the loop-carried dependencies, both at the op input and the body net input and output. (Also, did you mean op output?)

As a meta point, graph attributes are a general trait of the ONNX protobuf representation, so it might make sense to standardize these properties across all of ONNX, and not just have them part of the statement here.